Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.maven.project.MavenProject.getId()" because the return value of "org.apache.maven.project.ProjectBuildingResult.getProject()" is null

### Affected version

Maven 4.0.0-rc-4

### Bug description

I get this NPE in some projects when invoking mvn.
I can't really reproduce it, sometimes a mvn -U clean helps.
In my Gitlab pipeline this NPE is always there for at least two projects.
mvn -X -e gives
```
[ERROR] Internal error: java.lang.NullPointerException: Cannot invoke "org.apache.maven.project.MavenProject.getId()" because the return value of "org.apache.maven.project.ProjectBuildingResult.getProject()" is null -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.NullPointerException: Cannot invoke "org.apache.maven.project.MavenProject.getId()" because the return value of "org.apache.maven.project.ProjectBuildingResult.getProject()" is null
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:157)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:449)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:135)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke(Method.java:565)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.maven.project.MavenProject.getId()" because the return value of "org.apache.maven.project.ProjectBuildingResult.getProject()" is null
    at org.apache.maven.project.collector.DefaultProjectsSelector.selectProjects(DefaultProjectsSelector.java:78)
    at org.apache.maven.project.collector.RequestPomCollectionStrategy.collectProjects(RequestPomCollectionStrategy.java:49)
    at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor(DefaultGraphBuilder.java:367)
    at org.apache.maven.graph.DefaultGraphBuilder.build(DefaultGraphBuilder.java:100)
    at org.apache.maven.DefaultMaven.buildGraph(DefaultMaven.java:643)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:250)
    at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:225)
    at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:149)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.doExecute(MavenInvoker.java:449)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:97)
    at org.apache.maven.cling.invoker.mvn.MavenInvoker.execute(MavenInvoker.java:81)
    at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:165)
    at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:135)
    at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
    at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke(Method.java:565)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)```

--- a/impl/maven-core/src/main/java/org/apache/maven/internal/aether/TypeRegistryAdapter.java
+++ b/impl/maven-core/src/main/java/org/apache/maven/internal/aether/TypeRegistryAdapter.java
@@ -40,15 +40,12 @@ class TypeRegistryAdapter implements ArtifactTypeRegistry {
         if (type instanceof ArtifactType artifactType) {
             return artifactType;
         }
-        if (type != null) {
-            return new DefaultType(
-                    type.id(),
-                    type.getLanguage(),
-                    type.getExtension(),
-                    type.getClassifier(),
-                    type.isIncludesDependencies(),
-                    type.getPathTypes().toArray(new PathType[0]));
-        }
-        return null;
+        return new DefaultType(
+                type.id(),
+                type.getLanguage(),
+                type.getExtension(),
+                type.getClassifier(),
+                type.isIncludesDependencies(),
+                type.getPathTypes().toArray(new PathType[0]));
     }
 }

--- a/impl/maven-core/src/main/java/org/apache/maven/project/collector/DefaultProjectsSelector.java
+++ b/impl/maven-core/src/main/java/org/apache/maven/project/collector/DefaultProjectsSelector.java
@@ -75,7 +75,7 @@ public class DefaultProjectsSelector implements ProjectsSelector {
                         "{} {} encountered while building the effective model for '{}' (use -e to see details)",
                         problemsCount,
                         (problemsCount == 1) ? "problem was" : "problems were",
-                        result.getProject().getId());
+                        result.getProjectId());
 
                 if (request.isShowErrors()) { // this means -e or -X (as -X enables -e as well)
                     for (ModelProblem problem : result.getProblems()) {

--- a/impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultDependencyResolver.java
+++ b/impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultDependencyResolver.java
@@ -174,7 +174,8 @@ public class DefaultDependencyResolver implements DependencyResolver {
                         session.getNode(result.getRoot(), request.getVerbose()),
                         0);
             } catch (DependencyCollectionException e) {
-                throw new DependencyResolverException("Unable to collect dependencies", e);
+                String enhancedMessage = enhanceCollectionError(e, collectRequest);
+                throw new DependencyResolverException(enhancedMessage, e);
             }
         } finally {
             RequestTraceHelper.exit(trace);
@@ -273,5 +274,62 @@ public class DefaultDependencyResolver implements DependencyResolver {
 
     private static DependencyResolverException cannotReadModuleInfo(final Path path, final IOException cause) {
         return new DependencyResolverException("Cannot read module information of " + path, cause);
+    }
+
+    private static boolean containsUnresolvedExpression(String value) {
+        return value != null && value.contains("${") && value.contains("}");
+    }
+
+    private static String enhanceCollectionError(DependencyCollectionException e, CollectRequest request) {
+        if (e.getMessage() != null && e.getMessage().contains("Invalid Collect Request")) {
+            StringBuilder enhanced = new StringBuilder();
+            enhanced.append("Failed to collect dependencies");
+
+            org.eclipse.aether.graph.Dependency root = request.getRoot();
+            if (root != null && root.getArtifact() != null) {
+                org.eclipse.aether.artifact.Artifact artifact = root.getArtifact();
+                String groupId = artifact.getGroupId();
+                String artifactId = artifact.getArtifactId();
+                String version = artifact.getVersion();
+
+                if (containsUnresolvedExpression(groupId)
+                        || containsUnresolvedExpression(artifactId)
+                        || containsUnresolvedExpression(version)) {
+                    enhanced.append(" due to unresolved expression(s) in dependency: ")
+                            .append(groupId)
+                            .append(":")
+                            .append(artifactId)
+                            .append(":")
+                            .append(version)
+                            .append(".\n")
+                            .append("Please check that all properties are defined in your POM or settings.xml.");
+                    return enhanced.toString();
+                }
+            }
+
+            for (org.eclipse.aether.graph.Dependency dep : request.getDependencies()) {
+                if (dep != null && dep.getArtifact() != null) {
+                    org.eclipse.aether.artifact.Artifact artifact = dep.getArtifact();
+                    String groupId = artifact.getGroupId();
+                    String artifactId = artifact.getArtifactId();
+                    String version = artifact.getVersion();
+
+                    if (containsUnresolvedExpression(groupId)
+                            || containsUnresolvedExpression(artifactId)
+                            || containsUnresolvedExpression(version)) {
+                        enhanced.append(" due to unresolved expression(s) in dependency: ")
+                                .append(groupId)
+                                .append(":")
+                                .append(artifactId)
+                                .append(":")
+                                .append(version)
+                                .append(".\n")
+                                .append("Please check that all properties are defined in your POM or settings.xml.");
+                        return enhanced.toString();
+                    }
+                }
+            }
+        }
+        return e.getMessage();
     }
 }

--- a/its/core-it-suite/src/test/resources/mng-5208/spy/src/main/java/mng5208/EventLoggerSpy.java
+++ b/its/core-it-suite/src/test/resources/mng-5208/spy/src/main/java/mng5208/EventLoggerSpy.java
@@ -34,8 +34,7 @@ public class EventLoggerSpy extends AbstractEventSpy {
         if (event instanceof ExecutionEvent) {
 
             ExecutionEvent executionEvent = (ExecutionEvent) event;
-            System.out.println("executionEvent:" + executionEvent.getType() + "/"
-                    + executionEvent.getProject().getId());
+            System.out.println("executionEvent:" + executionEvent.getType() + "/" + executionEvent.getProjectId());
         }
     }
 }

